intel_adsp: move memory window to dts, add winstream console.

nashif opened the pull request:
Moving memory window to DTS. Memory windows were previously defined in multiple places (headers, kconfig, dts).
Partitioning is all defined per SoC now. 

Also move the winstream output to drivers/console and cleanup the boot code inside the soc/ layer.

This requires a new init level at the ARCH level and before we call c_start that can be implemented if we want to use drivers or services at an early stage of the boot process.


- kernel: init: introduce a new init level: ARCH
- intel_adsp: move memory windows to DTS proper
- console: move winstream console to drivers/

nashif committed:
intel_adsp: move memory windows to DTS proper

Instead of just declaring the memory window register in DTS and have
everything else all over the place (headers, Kconfig, etc.) this change
defines the memory window instances in DTS and uses the device model to
initialize the windows. Code is still part of the SoC, given that we do
not have a driver subsystem suitable for this type of device yet.

Move FW status to own workflow and separate from window setup.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
kernel: init: introduce a new init level: ARCH

We have cases where some devices needs to be initialized very early and
before c_start is call, i.e. to setup very early console or to setup
memory. Traditionally this would be hardcoded as part of the soc layer
and not using device model or the init levels.

This patch adds a new level ARCH, which will be called in early
architecture code and before we jump to the kernel code.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
console: move winstream console to drivers/

This was all done as part of the soc and called from the soc. Define
this type of console under drivers/console and use it in the SoCs
supporting that via SYS_INIT instead of calling the console code
directly.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

sylvioalves commented on soc/xtensa/intel_adsp/Kconfig.defconfig:
```diff
@@ -13,6 +13,14 @@ config XTENSA_CACHED_REGION
 config XTENSA_UNCACHED_REGION
 	default 4
 
+config CONSOLE
```
This should be placed after `if INTEL_ADSP_CAVS`, right?

andyross reviewed (commented):
This is for sure cleaner than what we have.  But really this memory is just plain SRAM.  It's not special, it doesn't need to be organized as part of hardware design.  Any (properly-aligned) address can be used.  So to my mind what we really want is to use the linker to pack these where they need to go, because the linker is the bit of software responsible for arranging bits of RAM.

So my personal ideal here would be kconfigs that specify size only, and then some code that just looks like:

```c
IF_ENABLED(CONFIG_INTEL_ADSP_WIN0_SIZE, \
           uint8_t __aligned(WHATEVER) win0_mem[CONFIG_INTEL_ADSP_WIN0_SIZE]));
```

That way if we decide in the future that we only need 128 bytes for the FW_STATUS window instead of 8k (yes, that's about the level of waste we're seeing right now!) it Just Works.  And if we have more complicated requirements in the future (e.g. we want IPC to be in a HP-SRAM bank that doesn't get powered down, but we can power down the trace and winstream windows, etc...) that becomes easy to express without having to manually move stuff around.

No reason not to merge something like this.  Just an area where I think we can do better still.

kv2019i commented on soc/xtensa/intel_adsp/common/mem_window.c:
```diff
@@ -0,0 +1,49 @@
+
+	sys_write32(config->size | 0x7, DMWLO(config->base_addr));
+	sys_write32((config->mem_base | CAVS_DMWBA_READONLY | CAVS_DMWBA_ENABLE),
+		    DMWBA(config->base_addr));
```
This is indeed much cleaner.

kv2019i commented on dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:
```diff
@@ -121,9 +121,38 @@
+			reg = <0x70210 0x8>;
+			memory = <&sram0>;
+			size = <0x2000>;
+			offset = <0x8000>;
```
For this we need some work on application (SOF) side to define DT overlays, as we two different schemes for the semantics of windows 2 and 3. We now use Kconfig overlays when building for different schemes with west. This mostly affects the cavs25 Intel target. For other ADSP targets we have in Zephyr, the expectation is that all apps will use the same definition (so the values in upstreadm Zephyr DT will be used by everybody).

kv2019i reviewed (commented):
Commenting on patch 2. Some work needed in SOF to take this into use, but I do agree this is cleaner to have the definitions in one place.

pjdobrowolski commented:
> This is for sure cleaner than what we have. But really this memory is just plain SRAM. It's not special, it doesn't need to be organized as part of hardware design. Any (properly-aligned) address can be used. So to my mind what we really want is to use the linker to pack these where they need to go, because the linker is the bit of software responsible for arranging bits of RAM.
> 
@andyross what if our FW is working in automotive industry where logging and data collecting is limited and we would like to collect data printed in memory window using some random reading tool not related to FW. Floating address of memory window might be very tedious to work with.

andyross commented:
> Floating address of memory window might be very tedious to work with.

For this particular hardware, the host-visible address is always constant.  Each window appears at a fixed offset in a PCI BAR no matter where it lives in the address space of the DSP.

But sure, you have to match the use case to the problem area.  Other hardware might be less capable, etc... But even then having the firmware provide some idea of a pointer/offset to the trace buffer or whatever seems pretty reasonable.

The problem I'm looking to solve here is that doing this manually has created a huge number of wasted bytes.  We reserve 64k on most of these platforms but are only using (vague numbers because I'm not the expert on most of this code) something like 184 bytes for FW_STATUS + IPC outbox, 256 bytes for IPC inbox, now 4k for @kv2019i's new mtrace facility, and if enabled 8k for Zephyr printk/logging.  That's a lot of memory that could be productively used for other things, but there's no way to do that easily now.  Make those into C arrays and everything just works.

kv2019i commented:
@andyross wrote:
> The problem I'm looking to solve here is that doing this manually has created a huge number of wasted bytes. We 
> reserve 64k on most of these platforms but are only using (vague numbers because I'm not the expert on most of this 
> code) something like 184 bytes for FW_STATUS + IPC outbox, 256 bytes for IPC inbox, now 4k for @kv2019i's new 

Did a bit of investigation around this and it turns out this is going to be pretty hard to do. The Linux driver actually queries the window offsets at boot, so it could handle "floating" window positions fairly easily, but the offset addresses (and sizes) are hardcoded in other places so the FW needs to keep these fixed at least for Intel builds. The amount of waste is going to also reduce as we'll have more users of the window2+3 with the "IPC4 scheme" (mtrace was first user).

kv2019i reviewed (commented):
This seems workable. Need to do a test round on hw to ensure the offsets are all good.

nashif commented on soc/xtensa/intel_adsp/common/mem_window.c:
```diff
@@ -0,0 +1,56 @@
+	}
+
+	sys_write32(config->size | 0x7, DMWLO(config->base_addr));
+	sys_write32((config->mem_base | CAVS_DMWBA_READONLY | CAVS_DMWBA_ENABLE),
```
yeah, we can do that in DTS, will take a look.

nashif committed:
intel_adsp: mem_window: rename defines CAVS_ -> ADSP_

Fix namespacing of memory window defines. Not all platforms are CAVS
based anymore.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
intel_adsp: mem_window: support read-only flag

Some windows might need to be set as writtable, so add a flag read-only
to DTS bindings which is set to true for all windows right now. This can
be set to false where needed.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

kv2019i reviewed (commented):
I think this looks great now. I'd still like to see results on a on-hardware test. I tried to schedule one now using SOF CI, let's see if it goes through 
https://github.com/thesofproject/sof/pull/6310

kv2019i reviewed (dismissed):
Test run passed in SOF CI, so looks good to go.

nashif commented:
> Test run passed in SOF CI, so looks good to go.

great, thanks. Now I have to fix some of the issues we have in our CI ;-(

mbolivar-nordic commented on dts/bindings/memory-window/intel,adsp-mem-window.yaml:
```diff
@@ -0,0 +1,20 @@
+  reg:
+    required: true
+
+  initialize:
+    type: boolean
```
Can you add a description here? This is kind of a vague property name.

mbolivar-nordic commented on soc/xtensa/intel_adsp/common/mem_window.c:
```diff
@@ -0,0 +1,56 @@
+
+#define DT_DRV_COMPAT intel_adsp_mem_window
+
+#define WIN_SIZE(N)   (CONFIG_MEMORY_WIN_##N##_SIZE)
+#define WIN0_OFFSET DT_INST_PROP(0, offset)
+#define WIN1_OFFSET WIN0_OFFSET + WIN_SIZE(0)
+#define WIN2_OFFSET WIN1_OFFSET + WIN_SIZE(1)
+#define WIN3_OFFSET WIN2_OFFSET + WIN_SIZE(2)
```
This is probably misuse of the INST APIs. Did you read https://docs.zephyrproject.org/latest/build/dts/api/api.html#c.DT_INST? Instance 0 is not necessarily the same thing as mem_window0. I think you should probably be using DT_NODELABEL instead.

nashif replied:
I was suspecting this would be the case, was too good to be true. Need to rework a few things then. Thanks.

mbolivar-nordic commented on drivers/console/Kconfig:
```diff
@@ -358,4 +358,10 @@ config EFI_CONSOLE
+	bool "Use Winstream console"
+	depends on WINSTREAM
+	help
+	  Use winstream as a console.
```
If you're moving this out of the soc/ folder and into drivers/console, I think it is worth it to improve the help text here. I can't figure out what winstream is by googling the name. Is it something generic? SoC specific? Etc

marc-hb replied:
+1

I think the best description can for now be found here: https://github.com/zephyrproject-rtos/zephyr/blame/53ef68d4/include/zephyr/sys/winstream.h
```
 ** @brief Lockless shared memory byte stream IPC
 *
 * The sys_winstream utility implements a unidirectional byte stream
 * with simple read/write semantics on top of a memory region shared
 * by the writer and reader. ...
```

Maybe @andyross can provide the Kconfig? It's his baby :-)

nashif replied:
it depends on WINSTREAM which is well defined in its own entry, no need for googling :)
```
config WINSTREAM
        bool "Lockless shared memory window byte stream"
        help
          Winstream is a byte stream IPC for use in shared memory
          "windows", generally for transmit to non-Zephyr contexts that
          can't share Zephyr APIs or data structures.
```

will update.

mbolivar-nordic replied:
Ah. That's what I was missing. Thanks.

mbolivar-nordic reviewed (dismissed):
Nice cleanup! My -1 is until we can sort out whether this is an appropriate use of DT_INST. Improving the description of the initialize property is up to you. The console help text is not my area so I can't -1 there, but I think it's a good idea.

nashif committed:
intel_adsp: mem_window: do not depend on instance numbering from DT

This is not guaranteed to match the physical layout of the memory, so
get them individually based on node label.

For initialization, use bbzero.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
intel_adsp: move ipc command out of mem window setup

Do FW communication outside of memory window setup to accomodate for
additional IPC commands and headers.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
ipm: intel_adsp: use memory window from DTS

Use memory window setup from DTS.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

mbolivar-nordic commented on soc/xtensa/intel_adsp/common/mem_window.c:
```diff
@@ -46,16 +36,56 @@ __imr int mem_win_init(const struct device *dev)
+};
+DEVICE_DT_DEFINE(DT_NODELABEL(mem_window1), mem_win_init, NULL, NULL, &mem_win_config_1, ARCH,
+		 CONFIG_KERNEL_INIT_PRIORITY_DEFAULT, NULL);
+#endif
```
I think it would lead to more maintainable code if you adopted this approach instead: https://github.com/zephyrproject-rtos/zephyr/blob/1d879b93c59e8f1f48d86db5794600769c86facb/drivers/spi/spi_nrfx_spim.c#L617

mbolivar-nordic replied:
Sorry, that line was confusing. I'm not talking about using a Kconfig symbol. I'm talking about using an approach that's similar, like this excerpt from the same driver:

```c
#define SPI_NRFX_SPIM_DEFINE(n) ...
#if ...
SPI_NRFX_SPIM_DEFINE(0);
#endif

#if ...
SPI_NRFX_SPIM_DEFINE(1);
#endif

...
```

You can do something similar:

```c
#define WINSTREAM_NODE(n) DT_NODELABEL(mem_window##n)
#define WINSTREAM_DEFINE(n) ... { .base_addr = DT_REG_ADDR(WINSTREAM_NODE(n)), ... } ...
#if DT_NODE_HAS_STATUS(WINSTREAM_NODE(0), okay)
WINSTREAM_DEFINE(0)
#endif
#if DT_NODE_HAS_STATUS(WINSTREAM_NODE(1), okay)
WINSTREAM_DEFINE(1)
#endif
...
```

nashif replied:
agree, I had to get it working first and get the blocks to a unified state first, now I can do the above.

galak commented on dts/bindings/memory-window/intel,adsp-mem-window.yaml:
```diff
@@ -0,0 +1,25 @@
+
+  offset:
+    type: int
+    description: offset from memory base.
```
what's the intent of offset, why can't `memory` just start at offset?  Seems like mixing of describing of the memory region, vs intent of some section of the memory region.

nashif replied:
memory is defined in some other node, offset is used here as an override and to deal with different layouts on different plaforms, the offset is actually calculated based on Kconfig, because it depend on what the application wants to do.

teburd reviewed (dismissed):
It's nice that the windows are now described and setup with device tree

kv2019i reviewed (commented):
Looks good. A bit worried whether this is bisect-safe, but couldn't point to any obvious place bisect would break, so seems good. 

The change to ipm_cavs_host.c in last commit should probably be moved to a more appropriate commit. Otherwise looks ok to go.

softwarecki commented on dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:
```diff
@@ -157,9 +157,35 @@
+			compatible = "intel,adsp-mem-window";
+			reg = <0x70208 0x8>;
+			memory = <&sram0>;
+			read-only;
```
Missing status "okay"?

softwarecki commented on dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:
```diff
@@ -157,9 +157,35 @@
+			reg = <0x70210 0x8>;
+			memory = <&sram0>;
+			read-only;
+		};
```
same as above

kv2019i commented on soc/xtensa/intel_adsp/cavs/include/intel_icl_adsp/adsp_memory.h:
```diff
@@ -22,15 +22,6 @@
-#define HP_SRAM_WIN2_SIZE (CONFIG_ADSP_WIN3_OFFSET - CONFIG_ADSP_WIN2_OFFSET)
-#define HP_SRAM_WIN3_BASE (L2_SRAM_BASE + CONFIG_ADSP_WIN3_OFFSET)
-#define HP_SRAM_WIN3_SIZE 0x2000
-
```
@nashif So in SOF side, these definitions are traditionally done (for each platform) in  src/platform/PLATFORM/include/platform/lib/memory.h
As we need to support non-Zephyr SOF builds as well in SOF, we've kept these definitions in place for platforms like "tgl". The definitions are part of the contract between SOF host and FW, so it's not like FW can change these values at its leisure, there are essentially part of the ABI contract.

Now for SOF MTL, this only supports a Zephyr build, so @aborisovich et al have done the right thing and used the Zephyr definitions:
sof/src/platform/meteorlake/include/platform/lib/memory.h

So if you now remove the old definitions in adsp_memory.h, you'll get a build error in SOF app. I think the SOF code could be updated to lookup the numbers from DTS directly (and this can be done for MTL as it's Zephyr only anyways). But transition needs to be managed to avoid build breakages inbetween.

I don't think this there is a big risk for duplicated definitions for these. If a the values are modified, you'll also break the host (i.e. this is picked up by the test sets very quickly).

nashif replied:
yes, this can worked around this for now.

kv2019i reviewed (commented):
Looks good now with the caveat that removing HP_SRAM_WIN0_BASE et al from adsp_memory.h will break MTL build in current SOF main. Other platforms are not currently impacted as they redefine the windows without using adsp_memory.h.

edersondisouza commented on soc/xtensa/intel_adsp/common/trace_out.c:
```diff
@@ -40,9 +43,16 @@ int arch_printk_char_out(int c)
 }
 
-void soc_trace_init(void)
+int soc_trace_init(void)
```
Changing the return type seems not relevant in this patch - maybe do this only in the next one?

galak commented on include/zephyr/init.h:
```diff
@@ -21,13 +21,14 @@ extern "C" {
 
 #ifdef CONFIG_SMP
-#define _SYS_INIT_LEVEL_SMP		4
+#define _SYS_INIT_LEVEL_SMP		5
```
Define these in terms of the level before.  For example:

`#define _SYS_INIT_LEVEL_SMP (_SYS_INIT_LEVEL_APPLICATION + 1)`

galak reviewed (commented):
A few devicetree description oriented ideas:

* it seems like the actual window number may NOT be important.  If that's the case, we should just "allocate" windows as they are used/needed
* we should introduce a 'mem-window' reference similar to GPIO which describes any params of the window and is defined in the node that is using the window.
* So for winstream, we should introduce a new devicetree node and have it have a `mem-window` property.

This a rough ideas that I need to look a bit more at what the actual uses are looking like and how the actual registers are being encoded.

Will try and provide a more detailed example of what the devicetree "nodes" would look like.

galak commented:
> So for winstream, we should introduce a new devicetree node and have it have a `mem-window` property.

Similarly we should convert drivers/ipm/ipm_cavs_host.c - to be devicetree based.  Something like this to start with (before any change for `mem-window`), there is probably a better name:

```
diff --git a/dts/xtensa/intel/intel_adsp_cavs.dtsi b/dts/xtensa/intel/intel_adsp_cavs.dtsi
index 1e4484dd6f..df4fd9432f 100644
--- a/dts/xtensa/intel/intel_adsp_cavs.dtsi
+++ b/dts/xtensa/intel/intel_adsp_cavs.dtsi
@@ -70,4 +70,9 @@
                        status = "okay";
                };
        };
+
+       ipc {
+               compatible = "intel,adsp-cavs-ipc";
+               status = "okay";
+       }
 };
```

andyross commented:
Just to jump in: I worry this is getting way too thick.  This is really simple hardware, but note also that it's hardware that needs to be enabled VERY early in the boot process (before the kernel code is even copied into place).  Devicetree complexity is fine as long as it stays build-time, but we can't tolerate any runtime data interpretation at all.

galak commented:
> Just to jump in: I worry this is getting way too thick. This is really simple hardware, but note also that it's hardware that needs to be enabled VERY early in the boot process (before the kernel code is even copied into place). Devicetree complexity is fine as long as it stays build-time, but we can't tolerate any runtime data interpretation at all.

Curious, what drives this needing to be early in the boot process?

marc-hb commented:
> Curious, what drives this needing to be early in the boot process?

Because it's the only communication with the outside world available early?

andyross commented:
Actually I meant Zephyr, but it works both ways. :)

The windows are shared memory.  The way it works is that the DSP memory space is mapped into the host PCI registers in these windows.  On the DSP/Zephyr side, they're specified by address/length pairs in the registers under discussion.  To the host (once mapped) they show up in[1] BAR 4, with 256kb of space per window (and an offset of 512kb from the start because there are registers in the same BAR).

Anyway, window zero is initialized not by Zephyr but by the ROM code in the DSP, where the first 32 bits are a "FW_STATUS register"[1] used to flag different stages of the load.  The final stage ("loaded", essentially) is written by the Zephyr kernel once it starts running.  But because the ROM put that thing somewhere we can't control, we need to remap it to our own address before we use it.  (Note that this all happens before the data copy, so we're still running out of "IMR" memory and haven't copied the image or done the .bss clear, both of which would have the effect of clearing FW_STATUS and breaking the load).

And as it turns out there IS a timing issue: if you are banging the DSP windows from the host at exactly the right moment as the registers are being set, the hardware will latch up and hang the DSP until the whole system does a power cycle.   The SOF kernel loader is pretty loose with its timings and seems robust, but early versions of our python loader In fact would try to spin on the register and could exercise this pretty reliably until I stumbled on the magic sleep that "fixed" it.  (And FWIW: audio failures in chromebooks on DSP load are a detectable failure case where I live now, and I'm reasonably certain it's this condition even if it's probably too rare for anyone to ever know for sure).

Anyway, that's probably more than you wanted to know.  But there are even more dragons, because the load isn't a window copy; the ROM actually uses the HD Audio DMA hardware to load firmware, then calls out to the CSME to check a signature on it.  Both of those steps can fail even before Zephyr starts running.

Fun times, basically.  In theory I've fled that world now, though, so I only have to deal with it through github comments. :)

[1] Numbers only probably correct, but still in my brain because they were burned there by extensive trauma.

[2] Not actually a register.

mbolivar-nordic commented on dts/xtensa/intel/intel_adsp_cavs18.dtsi:
```diff
@@ -79,6 +79,7 @@
 			reg = <0x71a08 0x8>;
 			memory = <&sram0>;
 			read-only;
+			status = "okay";
```
Something is wrong, I think.

This change doesn't actually do anything as far as I can tell.

Note that a missing status property is treated as if it were "okay". So you would have to explicitly change "disabled" to "okay" to have an affect. Adding a missing "okay" does nothing.

mbolivar-nordic replied:
> Note that a missing status property is treated as if it were "okay".

https://docs.zephyrproject.org/latest/build/dts/intro.html#important-properties

nashif replied:
> Something is wrong, I think.

it is okay, I just wanted to have them all enabled in a consistent way so it does not appear as if some are enabled and others are not. I could remove this alltogather and use the default for simplicity I guess, but the way it is now should be good as well.

nashif replied:
agree, let me remove that and use defaults instead.

mbolivar-nordic reviewed (commented):
The status patch looks fishy; did you verify it actually changes any behavior?

mbolivar-nordic commented on soc/xtensa/intel_adsp/common/mem_window.c:
```diff
@@ -36,56 +36,27 @@ __imr int mem_win_init(const struct device *dev)
-};
-DEVICE_DT_DEFINE(DT_NODELABEL(mem_window1), mem_win_init, NULL, NULL, &mem_win_config_1, ARCH,
-		 CONFIG_KERNEL_INIT_PRIORITY_DEFAULT, NULL);
+#if DT_NODE_HAS_STATUS(MEM_WINDOW_NODE(1), okay)
+MEM_WINDOW_DEFINE(1)
```
Should be squashed into the previous commit. Otherwise from a DT/driver side this all looks very reasonable to me.

kv2019i commented on dts/xtensa/intel/intel_adsp_cavs25.dtsi:
```diff
@@ -108,6 +109,7 @@
 			compatible = "intel,adsp-mem-window";
 			reg = <0x71a18 0x8>;
 			memory = <&sram0>;
+			read-only;
```
Hmm, the readonly definitions don't match SOF now:
https://github.com/thesofproject/sof/blob/main/src/platform/intel/cavs/lib/mem_window.c#L32

For all cAVS platforms, we have windows 0 and 3 as read-only and I believe same should apply for ace15.

FYI to @softwarecki who made comments about the settings earlier.

nashif replied:
> FYI to @softwarecki who made comments about the settings earlier.

that was already addressed.

kv2019i reviewed (changes requested):
Still one comment about the read-only settings for windows. See inline. 
Rest of the series seems good to go.

kv2019i reviewed (approved):
Series good to go!

gmarull commented on include/zephyr/init.h:
```diff
@@ -21,13 +21,14 @@ extern "C" {
-#define _SYS_INIT_LEVEL_PRE_KERNEL_2	1
-#define _SYS_INIT_LEVEL_POST_KERNEL	2
-#define _SYS_INIT_LEVEL_APPLICATION	3
+#define _SYS_INIT_LEVEL_ARCH		0
```
`device_service.c` missed the update

gmarull replied:
(and docs)

gmarull commented on tests/kernel/threads/no-multithreading/src/main.c:
```diff
@@ -95,7 +95,7 @@ FOR_EACH(SYS_INIT_CREATE, (;), PRE_KERNEL_1, PRE_KERNEL_2, POST_KERNEL);
 ZTEST(no_multithreading, test_sys_init)
 {
-	zassert_equal(init_order, 3, "SYS_INIT failed");
+	zassert_equal(init_order, _SYS_INIT_LEVEL_PRE_KERNEL_2, "SYS_INIT failed: %d", init_order);
```
this is wrong, it's a counter and must result in being incremented exactly 3 times, see L94.

gmarull reviewed (commented):
a couple of fixes are needed, reviewed too late.
